Dummy proposal PR (#1)
m-redding/azure-sdk-for-net: sdk/eventhub/Azure.Messaging.EventHubs/StreamingProducerProposal.md

    MaximumWaitTime = TimeSpan.FromMilliseconds(250),
    MaximumQueuedEventCount = 500
    RetryOptions = new EventHubsRetryOptions { TryTimeout = TimeSpan.FromMinutes(5) }
});    
```

### Creating the client with a fully qualified namespace and credential

```csharp
TokenCredential credential = new DefaultAzureCredential();
var fullyQualifiedNamespace = "<< NAMESPACE (likely similar to {your-namespace}.eventhub.windows.net) >>";
var eventHubName = "<< NAME OF THE EVENT HUB >>";

var producer = new StreamingProducer(fullyQualifiedNamespace, eventHubName, credential);
```

### Publish events using the Streaming Producer

```csharp
// Define handlers for the streaming events
    
protected async Task PublishSuccessfulHandlerAsync(PublishSucceededEventArgs args)
{
    foreach (var eventData in args.Events)
    {
        var eventId = eventData.Properties["event-id"];
        var partitionId = args.PartitionId
        Console.WriteLine($"Event: { eventId } was published by partition { partitionId }.");
    }
}

protected async Task PublishFailedHandlerAsync(PublishFailedEventArgs args)
{
    await Task.Delay(TimeSpan.FromSeconds(5));
    args.RetryPublishing = true;

> }

Reviewer: We may want to make this a bit more "real world" so that we don't get stuck in an infinite retry loop...

```c#
protected async Task PublishFailedHandlerAsync(PublishFailedEventArgs args)
{
    // If this exception is considered fatal, report it as a failure
    // and take no further action.

    if (IsFatalException(args.Exception, ...))
    {
        ReportFailure(args.Events, args.Exception, ...);
        return;
    }

    // Inspect the first event of the failed batch and determine how many 
    // additional retries we've used.
   
    var firstEvent = args.Events.First();

    var failedCount = firstEvent.Properties.TryGetValue("FailedTries", out var failCountObject) switch
    {
        true => ((int)failCountObject) + 1,
         _ => 1
    };

    // Allow for additional retries, above and beyond the built-in, after 
    // a delay for non-fatal exceptions.  If we've performed enough that
    // forward progress isn't being made, report the batch as a failure.
    
    if (failedCount <= 10)
    {
        // Track the failure tries on the first event of the batch, since we're
        // sure that this event will be included in the retry attempt batch.

        firstEvent.Properties["FailedTries"] = failedCount;

        await Task.Delay(TimeSpan.FromSeconds(5));        
        args.RetryPublishing = true;  
    }
    else
    {
        ReportFailure(args.Events, args.Exception, ...);
    }
}
```

    {
        foreach (var eventData in args.Events)
        {
            var eventId = eventData.Properties["event-id"];
            Console.WriteLine($"\t\t{ eventId } was too large to send");
            ReportPoisonEvent(eventData);
        }
    } 
    else 
    {    
        await Task.Delay(TimeSpan.FromSeconds(5));
        args.RetryPublishing = true;
    }
}

var connectionString = "<< CONNECTION STRING >>";
var eventHubName = "<< EVENT HUB NAME >>";

// Create the streaming producer with default options

var producer = new StreamingProducer(connectionString, eventHubName);

producer.PublishSucceeded += PublishSuccessfulHandlerAsync;    
producer.PublishFailed += PublishFailedHandlerAsync;

await producer.StartPublishingAsync();

try
{
    while (TryGetNextEvent(out var eventData))
    {
        await producer.QueueForSendAsync(eventData);
        Console.WriteLine($"There are { producer.TotalQueuedEventCount } events queued for publishing.");
    }
}
finally
{
    await producer.StopPublishingAsync();
    producer.PublishSucceeded -= PublishSuccessfulHandlerAsync;
    producer.PublishFailed -= PublishFailedHandlerAsync;
    
    await client.DisposeAsync();
}
```

> ### Checking that all messages are send successfully

Reviewer: Maybe we should consider calling this something like "Forcing queued events to be published"  or similar?

        Console.WriteLine($"Event: { eventId } was published by partition { partitionId }.");
    }
}

protected async Task PublishFailedHandlerAsync(PublishFailedEventArgs args)
{
    await Task.Delay(TimeSpan.FromSeconds(5));
    args.RetryPublishing = true;
}

var connectionString = "<< CONNECTION STRING >>";
var eventHubName = "<< EVENT HUB NAME >>";

// Create the streaming producer with default options

var producer = new StreamingProducer(connectionString, eventHubName);

producer.PublishSucceeded += PublishSuccessfulHandlerAsync;    
producer.PublishFailed += PublishFailedHandlerAsync;
await producer.StartPublishingAsync();

try
{
    while (TryGetNextEvent(out var eventData))
    {
        await producer.QueueForSendAsync(eventData);
        Console.WriteLine($"There are { producer.TotalQueuedEventCount } events queued for publishing.");
    }
    await producer.FlushAsync(); 
    Assert.AreEqual(0, producer.TotalQueuedEventCount);
}
finally
{
    await producer.StopPublishingAsync();
    producer.PublishSucceeded -= PublishSuccessfulHandlerAsync;
    producer.PublishFailed -= PublishFailedHandlerAsync;
    
    await client.DisposeAsync();
}
```

> ## Usage examples for potential additional features
> ### Creating client to ignore event ordering

Reviewer: ```suggestion
## Usage examples for potential additional features

### Creating client to ignore event ordering

```

When the ignore ordering flag is turned on then the order in which events are queued doesn't matter. By ignoring event ordering the producer can build and publish many batches at the same time since it doesn't matter if batch #1 fails and batch #2 succeeds because it can just resend number #1 without worrying about the order of delivery. This is useful when events are independent and can be processed in parallel without worrying about which was processed first. This functionality cannot be used at the same time as idempotency however, since idempotency depends on event ordering. 

> ```csharp 

Reviewer: ```suggestion

```csharp 
```

var connectionString = "<< CONNECTION STRING >>";
var eventHubName = "<< EVENT HUB NAME >>";

// Create the streaming producer

var producer = new StreamingProducer(connectionString, eventHubName, new StreamingProducerOptions
{
    AllowConcurrentPublishing = true; 
});    
```

> ### Publishing events using synchronous queueing 

Reviewer: ```suggestion
### Publishing events using synchronous queueing 

```

In the asynchronous `QueueForSendAsync()` method, in the case where the queue is full the method waits until there is space and then adds the event to the queue. In the synchronous `QueueForSend()` method, instead of hanging the method will immediately return and throw an error if the queue was full and queueing was unsuccessful. This allows the application to do something else while waiting for events on the queue to be published. One use case of this is to try to publish and then wait a specific amount of time before trying again, rather than continuously trying again and utilizing a lot of resources trying to do so.

> ```csharp

Reviewer: ```suggestion

```csharp
```

}
```
If the user would like to stop trying to publish and do something else rather than continuing to try to publish, then using `TryQueueForSend` is the recommended approach. This method attempts to add the events to the queue, and if the queue is full it just returns, rather than throwing an error.
```csharp
// Define handlers for the streaming events
    
protected async Task PublishSuccessfulHandlerAsync(PublishSucceededEventArgs args)
{
    foreach (var eventData in args.Events)
    {
        var eventId = eventData.Properties["event-id"];
        Console.WriteLine($"Event: { eventId } was published.");
    }
}

protected async Task PublishFailedHandlerAsync(PublishFailedEventArgs args)
{
    await Task.Delay(TimeSpan.FromSeconds(5));
    args.RetryPublishing = true;
}

// Create the producer client

var connectionString = "<< CONNECTION STRING >>";
var eventHubName = "<< EVENT HUB NAME >>";
var producer = new StreamingProducer(connectionString, eventHubName); 

producer.PublishSucceeded += PublishSuccessfulHandlerAsync;    
producer.PublishFailed += PublishFailedHandlerAsync;

await producer.StartPublishingAsync();

try
{
    TryGetNextEvent(out var eventData);

>     producer.TryQueueForSend(eventData);

Reviewer: We should flesh this out a bit more so that it reflects a real-world use.  Maybe something like:

```suggestion
    var queued = false;
    var attemptCount = 0;
    
    while ((!queued) && (++attemptCount < 10))
    {
        queued = producer.TryQueueForSend(eventData);
        
        if (!queued)
        {
            await Task.Delay(Timespan.FromSeconds(30));
        }
    }
    
    if (!queued)
    {
        throw new ApplicationException("Couldn't queue events to be published.");
    }
```

}
finally
{
    await producer.StopPublishingAsync();
    producer.PublishSucceeded -= PublishSuccessfulHandlerAsync;
    producer.PublishFailed -= PublishFailedHandlerAsync;
    
    await client.DisposeAsync();
}
```

> ### Using idempotent retries

Reviewer: Maybe we should consider removing this snippet and just including it in the skeleton?   I don't think it impacts the usage outside of when we validate options at construction.

}

public class PublishFailedEventArgs : EventArgs
{
    public IEnumerable<EventData> Events { get; init; }
    public Exception Exception { get; init; }
    public int Retries { get; init; }
    public int PartitionId { get; init ; }
    public PublishFailedEventArgs(IEnumerable<EventData> events, Exception ex, int partitionId);
}

public class StreamingProducerOptions : EventHubProducerClientOptions
{
    public TimeSpan? MaximumWaitTime { get; set; }
    public int MaximumQueuedEventCount { get; set; }   
    public boolean EnableIdempotentRetries { get; set; }   
    public boolean AllowConcurrentPublishing { get; set; }
}

public class StreamingProducer : IAsyncDisposable
{
    public event Action<StreamingProducer, PublishSucceededEventArgs> PublishSucceeded;
    public event Action<StreamingProducer, PublishFailedEventArgs> PublishFailed;

    public string FullyQualifiedNamespace { get; }
    public string EventHubName { get; }
    public string Identifier { get; }
    public int TotalQueuedEventCount { get; }
    public int GetPartitionQueuedEventCount(string partition = default);
    
    public bool IsRunning { get; protected set; }

>     public bool IsClosed { get; protected set; }

Reviewer: If we end up removing the disposable, we'll be able to kill this too.

    public StreamingProducer(string fullyQualifiedNamespace, string eventHubName, AzureSasCredential credential, StreamingProducerOptions streamingOptions = default);
    public StreamingProducer(string fullyQualifiedNamespace, string eventHubName, TokenCredential credential, StreamingProducerOptions streamingOptions = default);
    public StreamingProducer(EventHubConnection connection, EventHubProducerClientOptions clientOptions = default);

    public virtual Task StartPublishingAsync();
    
    // The queue is bounded and has limited capacity to prevent unchecked 
    // growth.  When the queue is full, this call waits for room.
    // 
    // Publishing will take place in the background, transparent to the caller.
    public virtual Task QueueForSendAsync(EventData eventData, CancellationToken cancellationToken, SendEventOptions options = default);
    
    // The queue is bounded and has limited capacity to prevent unchecked 
    // growth.  When the queue is full, this call throws an exception.
    // 
    // Publishing will take place in the background, transparent to the caller.

>     public virtual void QueueForSend(EventData eventData, CancellationToken cancellationToken, SendEventOptions options = default);

Reviewer: ```suggestion
```

I think we'll want to do the `Try` deal, so we should probably highlight that and remove this to avoid confusion.    _(we can adjust as feedback comes in)_

    public virtual void TryQueueForSend(EventData eventData, CancellationToken cancellationToken, SendEventOptions options = default);
    public virtual Task FlushAsync(CancellationToken cancellationToken);

    // Closing the producer will flush the queue, publishing any events that are still pending.
    public virtual Task StopPublishingAsync(CancellationToken cancellationToken);
    
    protected virtual void OnPublishSucceeded(IEnumerable<EventData> events);
    protected virtual void OnPublishFailed(IEnumerable<EventData> events, Exception ex, int partitionId);
}
```

> ## Competitive Analysis: Kafka 

Reviewer: ```suggestion
## Competitive Analysis: Kafka 

```

Reviewer: Love the side-by-side approach here;  we seem to come out looking good by comparison!

```java
// Creating the producer with basic properties
private Properties kafkaProperties = new Properties(); 
kafkaProperties.put("bootstrap.servers", "broker1:9092,broker2:9092");
kafkaProperties.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer"); 
kafkaProperties.put("value.serializer", "org.apache.kafka.common.serialization.StringSerializer");
producer = new KafkaProducer<String, String>(kafkaProperties);

// Sending a simple record through Kafka's send
ProducerRecord<String, String> record = new ProducerRecord<>("Topic", "Key", "Value"); 
try {
    producer.send(record); 
} catch (Exception e) {
    e.printStackTrace(); 
}
```

#### Sending a single message with streaming producer
The process from above is mimicked below using the Event Hubs streaming producer instead.
```csharp
// Create the producer client
var connectionString = "<< CONNECTION STRING >>";
var eventHubName = "<< EVENT HUB NAME >>";
var client = new EventHubProducerClient(connectionString, eventHubName)

// Create the streaming producer with default options
var producer = new StreamingProducer(client);

try
{

>     byte[] valuebytes = Encoding.ASCII.GetBytes("Value"); 
>     var eventData = EventData(valuebytes);

Reviewer: ```suggestion
   var eventData = EventData("Value");
```

New overload going in that will streamline this.  😄

var producer = new StreamingProducer(connectionString, eventHubName, new StreamingProducerOptions
{
    PartitionId = "0"
});
```

### Creating the client for a partition key

```csharp
var connectionString = "<< CONNECTION STRING >>";
var eventHubName = "<< EVENT HUB NAME >>";

// Create the streaming producer

var producer = new StreamingProducer(connectionString, eventHubName, new StreamingProducerOptions
{
    PartitionKey = "some_key"
});
```

### Creating the client with custom options

```csharp  
var connectionString = "<< CONNECTION STRING >>";
var eventHubName = "<< EVENT HUB NAME >>";

// Create the streaming producer

var producer = new StreamingProducer(connectionString, eventHubName, new StreamingProducerOptions
{
    Identifier = "Streaming Producer for Partition 0",
    PartitionId = "0",
    MaximumWaitTime = TimeSpan.FromMilliseconds(250),
    MaximumQueuedEventCount = 500
}, new EventHubProducerClientOptions
{
    EnableIdempotentPartitions = true,
    RetryOptions = new EventHubsRetryOptions { TryTimeout = TimeSpan.FromMinutes(5) }
});    
```

### Publish events using the Streaming Producer

```csharp
// Define handlers for the streaming events
    

> void PublishSuccessfulHandler(StreamingProducer sender, PublishSucceededEventArgs args)

Reviewer: We'll probably want these to be `async` to match the processor patterns.   Likewise, we can probably drop the `sender` parameter like we did there.

{

>     foreach (var eventData in args.Events)

Reviewer: We'll have multiple parallel operations happening if we've got more than one partition in use.... we should demonstrate that as well.   One of the `args` properties will likely need to be a partition id

    {
        var eventId = eventData.Properties["event-id"];
        Console.WriteLine($"Event: { eventId } was published.");
    }
}

void PublishFailedHandler(StreamingProducer sender, PublishFailedEventArgs args)
{
    Console.WriteLine("Publishing FAILED!");        
    Console.WriteLine("\tEvents:");
    
    foreach (var eventData in args.Events)
    {
        var eventId = eventData.Properties["event-id"];
        Console.WriteLine($"\t\t{ eventId } ");
    }
    
    Console.WriteLine($"\tException: { args.Exception.ToString() }");
}

var connectionString = "<< CONNECTION STRING >>";
var eventHubName = "<< EVENT HUB NAME >>";

// Create the streaming producer with default options

var producer = new StreamingProducer(connectionString, eventHubName);

producer.PublishSucceeded += PublishSuccessfulHandler;    
producer.PublishFailed += PublishFailedHandler;
await producer.StartProducerAsync();

try
{
    while (TryGetNextEvent(out var eventData))
    {
        await producer.QueueForSendAsync(eventData);

>         Console.WriteLine($"There are { producer.QueuedEventCount } events queued for publishing.");

Reviewer: This may get interesting;  if we're running separate queues for each partition and the gateway (automatic routing and partition key) we'll likely need a way to query the count with some context.    Any thoughts around what that may look like?

    }
}
finally
{

>     await producer.CloseAsync();

Reviewer: should we consider `StopPublishingAsync` for symmetry with the "start" call?

    producer.PublishSucceeded -= PublishSuccessfulHandler;
    producer.PublishFailed -= PublishFailedHandler;
    
    await client.DisposeAsync();
}

> ```

Reviewer: Should we consider demonstrating failure recovery by allowing immediate retry first?   Then we could use this as a branch of that logic.    There's not much real-world value in logging failures and moving on;  I suspect that we'll get questions about that upfront.    Maybe we show the basic success handler with an empty exception handler in the first snippet or two and then demonstrate the common path of "please just keep retrying" here, followed by the "well, I don't want to retry."

Thoughts?

> ### Failure recovery: Message Size Exceeded

Reviewer: ```suggestion
### Failure recovery: poison messages

```

> When publishing to the Event Hub occasionally an error may come up, either a connectivity issue, an issue with the client itself, or something else. There are many ways to deal with failures depending on what the goal of the application is, but this is one way that involves catching a message size exceeded exception, meaning that the message being sent does not fit in the batch and will never fit in the batch.

Reviewer: ```suggestion
When publishing to the Event Hub occasionally an error that is not resolved through retires may occur.  While some may be recovered if allowed to continue retrying, others may be terminal.  For example, a poison message that will never succeed.  For Event Hubs, this is most commonly the result of an event exceeding the maximum message size allowed by the Event Hubs service.

```

```csharp
// Define handlers for the streaming events
    
void PublishSuccessfulHandler(StreamingProducer sender, PublishSucceededEventArgs args)
{
    foreach (var eventData in args.Events)
    {
        var eventId = eventData.Properties["event-id"];
        Console.WriteLine($"Event: { eventId } was published.");
    }
}

void PublishFailedHandler(StreamingProducer sender, PublishFailedEventArgs args)
{
    Console.WriteLine("Publishing FAILED!");   

>     if (args.Exception.Reason == EventHubsException.FailureReason.MessageSizeExceeded)

Reviewer: What happens if this is not an `EventHubsException`?   We'd fail on this line.

    {
        foreach (var eventData in args.Events)
        {
            var eventId = eventData.Properties["event-id"];
            Console.WriteLine($"\t\t{ eventId } was too large to send");

>             LogEvent(eventData);

Reviewer: We're mixing `Console.WriteLine` with our fake `LogEvent`  method, which may be confusing.   Maybe we should consider making this something like `ReportPoisonEvent` or similar so that the context is clear at a glance?

        }

>     } else {    

Reviewer: nit:  Allman braces, we're not monsters!   😄 

```suggestion
    } 
    else 
    {    
```

        Console.WriteLine("\tEvents:");
        
        foreach (var eventData in args.Events)
        {
            var eventId = eventData.Properties["event-id"];
            Console.WriteLine($"\t\t{ eventId } ");
        }
        
        Console.WriteLine($"\tException: { args.Exception.ToString() }");
    }
}

var connectionString = "<< CONNECTION STRING >>";
var eventHubName = "<< EVENT HUB NAME >>";

// Create the streaming producer with default options

var producer = new StreamingProducer(connectionString, eventHubName);

producer.PublishSucceeded += PublishSuccessfulHandler;    
producer.PublishFailed += PublishFailedHandler;

await producer.StartProducerAsync();

try
{
    while (TryGetNextEvent(out var eventData))
    {
        await producer.QueueForSendAsync(eventData);
        Console.WriteLine($"There are { producer.QueuedEventCount } events queued for publishing.");
    }
}
finally
{
    await producer.CloseAsync();
    producer.PublishSucceeded -= PublishSuccessfulHandler;
    producer.PublishFailed -= PublishFailedHandler;
    
    await client.DisposeAsync();
}
```

> ### Failure recovery: Resource not Found

Reviewer: I'm not sure this shows us anything more than the poison message scenario.   For the immediate retry we'd need to signal the producer by working with the args.   Something like:

```csharp
await Task.Delay(TimeSpan.FromSeconds(5));
args.RetryPublishing = true;
```

In this case, I'm not sure how a resource would be found after a delay, though.   That may be better demonstrated with an `IOException` or `EventHubsExxception` with reason `ServiceCommunicationFailure` or  similar.

# Streaming Producer: Proposal

> Publishing events using the current producer works very well for high and consistent throughput, and allows applications to treat sets of events as a one unit. Applications send batches through the producer and the producer returns whether the publishing was successful. The client can send events to any partition and can publish events concurrently.

Reviewer: ```suggestion
Publishing events using the producer client is optimized for high and consistent throughput scenarios, allowing applications to collect a set of events as a batch and publish in a single operation.  In order to maximize flexibility, developers are expected to built and manage batches according to the needs of their application, allowing them to prioritize trade-offs between ensuring batch density, enforcing strict ordering of events, and publishing on a consistent and predictable schedule.
```

> The goal of the streaming producer is to provide applications with the ability to send individual events without needing to implement the overhead of collecting events, constructing batches, and then publishing them. Events are sent and published in the order they were queued, and which events go into a batch cannot be controlled by the application itself. The application can dynamically choose whether to send each event to a specific partition or through the main application gateway. 

Reviewer: ```suggestion
The primary goal of the streaming producer is to provide developers the ability to queue individual events for publishing without the need to explicitly manage batch construction, population, or service operations.  Events are collected as they are queued, organized into batches, and published by the streaming producer as batches become full or a certain amount of time has elapsed.  When queuing events, developers may request automatic routing to a partition or explicitly control the partition in the same manner supported by the `ProducerClient`, with the streaming producer managing the details of grouping events into the appropriate batches for publication.

An additional goal of the streaming producer is to support idempotent retries in a way that is intuitive, helping to reduce the potential for duplicate events to be published in many common failure scenarios.
```

> An additional goal of the streaming producer is to provide idempotency in a way that is intuitive and efficient. A flag may be used to signal idempotent event handling in this streaming producer, rather than in the main client producer as previously investigated. 

Reviewer: ```suggestion
```

        Console.WriteLine($"Event: { eventId } was published.");
    }
}

void PublishFailedHandler(StreamingProducer sender, PublishFailedEventArgs args)
{
    Console.WriteLine("Publishing FAILED!");        
    Console.WriteLine("\tEvents:");
    
    foreach (var eventData in args.Events)
    {
        var eventId = eventData.Properties["event-id"];
        Console.WriteLine($"\t\t{ eventId } ");
    }
    
    Console.WriteLine($"\tException: { args.Exception.ToString() }");
}

// Create the producer client

var connectionString = "<< CONNECTION STRING >>";
var eventHubName = "<< EVENT HUB NAME >>";

var client = new EventHubProducerClient(connectionString, eventHubName)

// Create the streaming producer with retry option

var producer = new StreamingProducer(client, new StreamingProducerOptions
{
    Retries = int.MaxValue; 
}); 

producer.PublishSucceeded += PublishSuccessfulHandler;    
producer.PublishFailed += PublishFailedHandler;

await producer.StartProducerAsync();

var sendOptions = new SendEventOptions
{
    PartitionId = firstPartition
};
try
{
    while (TryGetNextEvent(out var eventData))
    {
        try {

>             producer.QueueForSend(eventData, sendOptions);

Reviewer: What happens if the queue is full here?    We should probably consider explicitly handling that.   The attempt to call again in the `catch` block that we're demonstrating is also likely to fail in the current form.   We may need to consider something like an inner loop:

```csharp
    while (TryGetNextEvent(out var eventData))
    {
        var queued = false;
        
        while (!queued )
        {
            try 
            {
                producer.QueueForSend(eventData, sendOptions);
                queued = true;
            }
            catch (SomeQueueFullException)
            {
                await Task.Delay(Timespan.FromSeconds(30));
            }
        }
    }
```

Reviewer: Seeing this.... I wonder.    Should we use a `TryQueueForSend` method instead?   Do we think that people will stop trying to publish if they see it blocked or are more likely to do a delay-retry pattern?

});

producer.PublishSucceeded += PublishSuccessfulHandler;    
producer.PublishFailed += PublishFailedHandler;
await producer.StartProducerAsync();

try
{
    while (TryGetNextEvent(out var eventData))
    {
        await producer.QueueForSendAsync(eventData);
        Console.WriteLine($"There are { producer.QueuedEventCount } events queued for publishing.");
    }
}
finally
{
    await producer.CloseAsync();
    producer.PublishSucceeded -= PublishSuccessfulHandler;
    producer.PublishFailed -= PublishFailedHandler;
    
    await client.DisposeAsync();
}
```

## API skeleton
This API skeleton includes the planned additional features as demonstrated above.
### `Azure.Messaging.EventHubs.Producer`

```csharp
public class PublishSucceededEventArgs : EventArgs
{
    public IEnumerable<EventData> Events { get; init; }
    public PublishSucceededEventArgs(IEnumerable<EventData> events);
}

public class PublishFailedEventArgs : EventArgs
{
    public IEnumerable<EventData> Events { get; init; }
    public Exception Exception { get; init; }
    public int Retries { get; init; }
    public PublishFailedEventArgs(IEnumerable<EventData> events, Exception ex, int numRetries = 0);
}

> public class StreamingProducerOptions
> {
>     public string Identifier { get; set; }
>     public string PartitionKey { get; set; }
>     public string PartitionId { get; set; }
>     public TimeSpan? MaximumWaitTime { get; set; }
>     public int MaximumQueuedEventCount { get; set; }   
>     public boolean IdempotentSend { get; set; }   
> }

Reviewer: ```suggestion
public class StreamingProducerOptions : EventHubProducerClientOptions
{
    public TimeSpan? MaximumWaitTime { get; set; }
    public int MaximumQueuedEventCount { get; set; }   
    public bool EnableIdempotentRetries { get; set; }   
}
```

This would give us the common core options for free as well as letting us create the producer without having to copy the values over.

> public class StreamingProducer

Reviewer: We'll probably want to consider making this implement `IAsyncDisposable` to match the other Event Hubs client types.   Though I consider that a mistake, the consistency would be good.

{
    public event Action<StreamingProducer, PublishSucceededEventArgs> PublishSucceeded;
    public event Action<StreamingProducer, PublishFailedEventArgs> PublishFailed;

    public string FullyQualifiedNamespace { get; }
    public string EventHubName { get; }
    public string Identifier { get; }

>     public int QueuedEventCount { get; }

Reviewer: How do we change this up to allow us to query the queued event count for a given partition or the gateway?   Is it as simple as something like:

```csharp
public int GetQueuedEventCount(string partition = default);
```

or do we have something that makes it more clear somehow?

>     public bool IsActive { get; protected set; }

Reviewer: We can probably make this `IsRunning` to match the processor, especially if we're doing the `Start` and `Stop` named methods

    public bool IsClosed { get; protected set; }
    

>     public StreamingProducer(string connectionString, string eventHubName, StreamingProducerOptions streamingOptions = default, EventHubProducerClientOptions clientOptions = default);

Reviewer: For the constructors, I believe we'd want to match the same set of overloads that the `EventHubProducerClient` offers, albeit with the streaming producer options type.

    protected StreamingProducer();

    public virtual Task StartProducerAsync();
    
    // The queue is bounded and has limited capacity to prevent unchecked 
    // growth.  When the queue is full, this call waits for room.
    // 
    // Publishing will take place in the background, transparent to the caller.
    public virtual Task QueueForSendAsync(EventData eventData, CancellationToken cancellationToken, SendEventOptions options = default);
    
    // The queue is bounded and has limited capacity to prevent unchecked 
    // growth.  When the queue is full, this call throws an exception.
    // 
    // Publishing will take place in the background, transparent to the caller.
    public virtual void QueueForSend(EventData eventData, CancellationToken cancellationToken, SendEventOptions options = default);

>     public virtual Task FlushAsync(CancellationToken cancellationToken);

Reviewer: ooohh.... nice.   I forgot about this.   We should probably demonstrate this in a snippet if I didn't overlook one already.

    // Closing the producer will flush the queue, publishing any events that are still pending.
    public virtual Task CloseAsync(CancellationToken cancellationToken);
    

>     protected virtual void OnPublishSucceeded(EventData eventData);

Reviewer: This would probably need to take `IEnumerable<EventData>` to match the pattern that we're demonstrating in the handler.

>     protected virtual void OnPublishFailed(EventData eventData, Exception ex);

Reviewer: We'll probably want to allow for some kind of context to detail the partition or partition key that the failure impacted, if one was in use.   Is there any other context that we think would be helpful for helping establish context?

}
```

## Competitive Analysis: Kafka 

> The streaming producer will be competitive with most of the features provided by Kafka's producer. Both allow events to be added one at a time to a queue of events that asynchronously send the events after batching them together. Both also have the following options: allowing retries or not, restricting the number of events held in the queue, set the timeout value for sending a partial batch, and turning on or off idempotent retries. 

Reviewer: ```suggestion
The streaming producer will offer parity with most of the features provided by Kafka's producer. Both allow events to be added one at a time to a queue of events which are asynchronously published after batching them together. Both also support the following options: allowing retries or not, restricting the number of events held in the queue, setting the timeout value for sending a partial batch, and enabling idempotent retries. 
```

In this context is "setting the timeout"  meant to indicate the time to flush after inactivity or the time to allow the publish operation to take place before considering it a failure?

> ### Coming from Kafka 
> #### Sending a single message with Kafka's producer

Reviewer: ```suggestion
### Coming from Kafka 

#### Sending a single message with Kafka's producer

```
